File: ch37p1099downloader/ch37p1099downloader/MyTableViewController.m


#import "MyTableViewController.h"
#import "MyDownloader.h"

@interface MyTableViewController () <UITableViewDelegate, UITableViewDataSource>
@property (nonatomic, strong) NSMutableArray* model;
@property (nonatomic, strong) MyDownloader* downloader;
@end

@implementation MyTableViewController

- (NSURLSessionConfiguration*) configuration {
    NSURLSessionConfiguration* config =
    [NSURLSessionConfiguration ephemeralSessionConfiguration];
    config.allowsCellularAccess = NO;
    config.URLCache = nil;
    return config;
}

- (void) viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    if (!self.downloader) {
        MyDownloader* downloader = [[MyDownloader alloc] initWithConfiguration:[self configuration]];
        self.downloader = downloader;
    }
    if (self.model)
        return;
    NSString* mannyurl = @"http://www.apeth.com/pep/manny.jpg";
    NSString* jackurl = @"http://www.apeth.com/pep/jack.jpg";
    NSString* moeurl = @"http://www.apeth.com/pep/moe.jpg";
    self.model = [NSMutableArray array];
    
    for (int ix = 0; ix < 15; ix++) {
        NSMutableDictionary* d = [NSMutableDictionary dictionaryWithDictionary:@{@"text": @"Manny", @"pic": mannyurl}];
        [self.model addObject: d];
    }
    for (int ix = 15; ix < 30; ix++) {
        NSMutableDictionary* d = [NSMutableDictionary dictionaryWithDictionary:@{@"text": @"Moe", @"pic": moeurl}];
        [self.model addObject: d];
    }
    for (int ix = 30; ix < 45; ix++) {
        NSMutableDictionary* d = [NSMutableDictionary dictionaryWithDictionary:@{@"text": @"Jack", @"pic": jackurl}];
        [self.model addObject: d];
    }
}

// Customize the number of sections in the table view.
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

// Customize the number of rows in the table view.
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (!self.model)
        return 0;
    return [self.model count];
}

// Customize the appearance of table view cells.
- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    NSMutableDictionary* d = (self.model)[indexPath.row];
    cell.textLabel.text = d[@"text"];
    // have we got a picture?
    id pic = d[@"pic"];
    if ([pic isKindOfClass:[UIImage class]]) {
        cell.imageView.image = pic;
    } else if ([pic isKindOfClass:[NSString class]]) {
        cell.imageView.image = nil;
        [self.downloader download:pic completionHandler:^(NSURL* url){
            if (!url)
                return;
            NSData* data = [NSData dataWithContentsOfURL:url];
            UIImage* im = [UIImage imageWithData:data];
            d[@"pic"] = im;
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.tableView reloadRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationNone];
            });
        }];
    }
    return cell;
}

- (void) dealloc {
    if (self->_downloader)
        [self->_downloader cancelAllTasks];
    NSLog(@"%@", @"table view controller dealloc");
}



@end

File: iOS12bookExamples/bk2ch24p842downloader/ch37p1099downloader/MyTableViewController.swift


import UIKit

class MyTableViewController: UITableViewController, UITableViewDataSourcePrefetching {

    lazy var configuration : URLSessionConfiguration = {
        let config = URLSessionConfiguration.ephemeral
        config.allowsCellularAccess = false
        config.urlCache = nil
        return config
        }()
    
    lazy var downloader : Downloader = {
        return Downloader(configuration:self.configuration)
    }()

    var model : [Model] = {
        let mannyurl = "https://www.apeth.com/pep/manny.jpg"
        let jackurl = "https://www.apeth.com/pep/jack.jpg"
        let moeurl = "https://www.apeth.com/pep/moe.jpg"
        var arr = [Model]()
        for _ in 0 ..< 15 {
            let m = Model()
            m.text = "Manny"
            m.picurl = mannyurl
            arr.append(m)
        }
        for _ in 0 ..< 15 {
            let m = Model()
            m.text = "Moe"
            m.picurl = moeurl
            arr.append(m)
        }
        for _ in 0 ..< 15 {
            let m = Model()
            m.text = "Jack"
            m.picurl = jackurl
            arr.append(m)
        }
        return arr
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.tableView.prefetchDataSource = self // turn on prefetching
    }
    
    var didSetup = false
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.model.count
    }
    
    func tableView(_ tableView: UITableView, prefetchRowsAt indexPaths: [IndexPath]) {
        for ip in indexPaths {
            let m = self.model[ip.row]
            guard m.im == nil else { print("nop \(ip)"); return } // we have an image, nothing to do
            guard m.task == nil else { print("nop2 \(ip)"); return } // we're already downloading
            print("prefetching for \(ip)")
            let url = URL(string:m.picurl)!
            m.task = self.downloader.download(url:url) { url in
                m.task = nil
                if let url = url, let data = try? Data(contentsOf: url) {
                    print("got \(ip)")
                    m.im = UIImage(data:data)
                    tableView.reloadRows(at:[ip], with: .none)
                }
            }
        }
    }
    
	let cellID = "Cell"
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: self.cellID, for: indexPath)
        let m = self.model[indexPath.row]
        cell.textLabel!.text = m.text
        cell.imageView!.image = m.im // image or nil
        if m.task == nil && m.im == nil {
            self.tableView(tableView, prefetchRowsAt:[indexPath])
        }
        return cell
    }
    
    // uncomment to try expunging
   /*
    override func tableView(_ tableView: UITableView, didEndDisplaying cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        let m = self.model[(indexPath as NSIndexPath).row]
        if m.task == nil && m.im != nil {
            m.im = nil // expunge
        }
    }
 */
    
    deinit {
        print("table view controller dealloc")
    }

}

// unfortunately a Swift dictionary inside an array is effectively immutable
// but we need to mutate our model objects
// one solution would be to resort to NSMutableDictionary
// but the truth is that this should have been a simple value class all along, so here it is

class Model {
    var text : String!
    var im : UIImage!
    var picurl : String!
    var task : URLSessionTask!
}
